File: python/863. All Nodes Distance K in Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def distanceK(self, root, target, K):
        """
        :type root: TreeNode
        :type target: TreeNode
        :type K: int
        :rtype: List[int]
        """
        def dfs(root, parent):
            if not root:
                return
            root.parent = parent
            dfs(root.left, root)
            dfs(root.right, root)
        
        dfs(root, None)

        K += 1
        visit = set()
        ans = []
        que = [target]
        while que:
            tmp = []
            K -= 1
            sz = len(que)
            while sz:
                sz -= 1
                root = que.pop(0)
                visit.add(root)
                tmp.append(root.val)
                if root.left and root.left not in visit:
                    que.append(root.left)
                if root.right and root.right not in visit:
                    que.append(root.right)
                if root.parent and root.parent not in visit:
                    que.append(root.parent)
            if K==0:
                ans = tmp
                break
        return ans
                

            

File: go/src/leetcode/863. All Nodes Distance K in Binary Tree/863. All Nodes Distance K in Binary Tree.go
package main

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

type Node struct {
	Val int
	Left *Node
	Right *Node
	Parent *Node
}

func distanceK(root *TreeNode, target *TreeNode, k int) []int {
	var t *Node

	var traverse func(node *TreeNode) *Node
	traverse = func(node *TreeNode) *Node {
		if node == nil {
			return nil
		}
		root := &Node{
			Val: node.Val,
			Left: traverse(node.Left),
			Right: traverse(node.Right),
		}
		if target == node {
			t = root
		}
		if root.Left != nil {
			root.Left.Parent = root
		}
		if root.Right != nil {
			root.Right.Parent = root
		}
		return root
	}

	traverse(root)

	vis := map[*Node]bool{}
	var ans = []int{}
	que := []*Node{t}
	vis[t] = true
	for ; len(que) > 0 && k >= 0; k-- {
		for sz := len(que); sz > 0; sz-- {
			t := que[0]
			que = que[1:]
			ans = append(ans, t.Val)
			if t.Left != nil && !vis[t.Left] {
				vis[t.Left] = true
				que = append(que, t.Left)
			}
			if t.Right != nil && !vis[t.Right] {
				vis[t.Right] = true
				que = append(que, t.Right)
			}
			if t.Parent != nil && !vis[t.Parent] {
				vis[t.Parent] = true
				que = append(que, t.Parent)
			}
		}
		ans = []int{}
	}
	return ans
}
